### PR TITLE
[#5447] Update authentication logic in hs_core/authentication.py

### LINKED ISSUE #5447: Return meaningful response when authenticate with inacurate credentials
**Description of the bug**
Authentication attempt using hsrest client with bad (typo etc) credentials.
```
---------------------------------------------------------------------------
HydroShareHTTPException                   Traceback (most recent call last)
<ipython-input-4-9e65448ed82e> in <module>
----> 1 hs.getResourceFile('bc14db4a50bd4d89bb750126a0a1ba79', file_name, destination=PROJECT_DIR)

/cvmfs/cybergis.illinois.edu/software/conda/cjw/pyrhessys-2024-02/lib/python3.7/site-packages/hs_restclient/__init__.py in getResourceFile(self, pid, filename, destination)
    901                 raise HydroShareNotFound((pid, filename))
    902             else:
--> 903                 raise HydroShareHTTPException(r)
    904 
    905         if destination is None:

HydroShareHTTPException: Received status 500 <!doctype html>
<html lang="en">
```
```
[06/Apr/2024 09:29:39] ERROR [django.request:224] Internal Server Error: /hsapi/resource/bc14db4a50bd4d89bb750126a0a1ba79/files/CoweetaSub18_10m_TEST.zip/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 497, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 414, in initial
    self.perform_authentication(request)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 324, in perform_authentication
    request.user
  File "/usr/local/lib/python3.9/site-packages/rest_framework/request.py", line 227, in user
    self._authenticate()
  File "/usr/local/lib/python3.9/site-packages/rest_framework/request.py", line 380, in _authenticate
    user_auth_tuple = authenticator.authenticate(self)
  File "/hydroshare/hs_core/authentication.py", line 96, in authenticate
    user = User.objects.get(username=decoded_username)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 435, in get
    raise self.model.DoesNotExist(
django.contrib.auth.models.User.DoesNotExist: User matching query does not exist.
```

Add exception handling to [query for user](https://github.com/hydroshare/hydroshare/blob/2.13.6/hs_core/authentication.py#L96) to return 403 or other appropriate responses instead of 500.

**Additional information**
hs V 2.13.6

### PR DESCRIPTION
**Pull request by devincowan**
Resolves #5447 

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### CODE REVIEW

File: hs_core/authentication.py [hunk 1]
```diff
@@ -92,6 +92,4 @@ def authenticate(self, request):
             KEYCLOAK.token(decoded_username, decoded_password)
         except Exception:
             return None
-
-        user = User.objects.get(username=decoded_username)
-        return (user, None)
```
Comment (devincowan) [inline hs_core/authentication.py:97]: not sure if there was a reason we were returning a tuple here...
https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#writing-an-authentication-backend

File: theme/backends.py [hunk 2]
```diff
@@ -40,17 +40,12 @@ class CaseInsensitiveMezzanineBackend(ModelBackend):
...
+            if username_or_email:
                 password = kwargs.pop("password", None)
-                try:
-                    user = User.objects.get(username_or_email, **kwargs)
-                except User.DoesNotExist:
-                    pass
-                else:
-                    if user.check_password(password):
-                        return user
+                user = get_user_from_username_or_email(username_or_email, **kwargs)
```
[the remark this answers is not in the document]
Comment (sblack-usu) [reply theme/backends.py:46]: Ah, I didn't see that. I doubt those kwargs are used but better to be safe.

File: hs_core/authentication.py (hunk 1 above)
Comment (sblack-usu) [reply hs_core/authentication.py:97]: The example I was looking at returned the tuple, it must've been an older example.  No matter, it is not needed in our current version.

File: theme/utils.py [hunk 3]
```diff
@@ -82,3 +83,19 @@ def get_quota_message(user):
                                                               zone=uq.zone,
                                                               percent=rounded_percent)
         return return_msg
+
+
+def get_user_from_username_or_email(username_or_email, **kwargs):
```
Comment (devincowan) [inline theme/utils.py:88]: @sblack-usu all I did was type "def get_user_from_username_or_email" and CoPilot wrote this function... :)

File: hs_core/authentication.py [hunk 4]
```diff
@@ -5,9 +5,9 @@
 from django.urls import reverse, resolve
 from django.conf import settings
 from django.utils.http import urlencode
-from django.contrib.auth.models import User
 from rest_framework.authentication import BaseAuthentication
 from keycloak.keycloak_openid import KeycloakOpenID
+from theme.utils import get_user_from_username_or_email
```
Comment (sblack-usu) [inline hs_core/authentication.py:10]: I'm having second thoughts about the utils idea. We currently don't have any imports to theme in hs_core (and we also don't have any imports of hs_core into theme).  I think I would rather have theme importing hs_core than the other way around. 

I'm also looking at the rest_framework import and realize that wasn't great of me to put this rest_framework authentication class in hs_core. I don't want this patch to be held up by refactoring and we need to get this patch out soon so.

I'll write a tech debt ticket to capture these concerns and to help inform a later refactor and approve this ticket.

**Review by sblack-usu — APPROVED** [review]

File: hs_core/authentication.py (hunk 4 above)
Comment (devincowan) [reply hs_core/authentication.py:10]: good call

**Pull request merged**